File: iOS/Services/NotificationReliability.swift
import Foundation

public enum NotificationReliabilityLevel: String, Sendable {
    case healthy
    case degraded
    case needsRepair
    case inactive
}

public struct NotificationReliabilityInput: Sendable, Equatable {
    public let permissionGranted: Bool
    public let enabledByUser: Bool
    public let hasActiveClient: Bool
    public let activeSupplementCount: Int
    public let pendingCount: Int
    public let pendingOnlyCount: Int
    public let shadowOnlyCount: Int
    public let shadowErrorCount: Int

    public init(
        permissionGranted: Bool,
        enabledByUser: Bool,
        hasActiveClient: Bool,
        activeSupplementCount: Int,
        pendingCount: Int,
        pendingOnlyCount: Int,
        shadowOnlyCount: Int,
        shadowErrorCount: Int
    ) {
        self.permissionGranted = permissionGranted
        self.enabledByUser = enabledByUser
        self.hasActiveClient = hasActiveClient
        self.activeSupplementCount = activeSupplementCount
        self.pendingCount = pendingCount
        self.pendingOnlyCount = pendingOnlyCount
        self.shadowOnlyCount = shadowOnlyCount
        self.shadowErrorCount = shadowErrorCount
    }
}

public struct NotificationReliabilityReport: Sendable, Equatable {
    public let level: NotificationReliabilityLevel
    public let mismatchCount: Int
    public let shouldOfferRepair: Bool
}

public enum NotificationReliabilityEvaluator {
    public static func evaluate(_ input: NotificationReliabilityInput) -> NotificationReliabilityReport {
        let inactive = !input.enabledByUser || !input.permissionGranted || !input.hasActiveClient
        if inactive || input.activeSupplementCount == 0 {
            return NotificationReliabilityReport(level: .inactive, mismatchCount: mismatchCount(input), shouldOfferRepair: false)
        }
        let mismatches = mismatchCount(input)
        if mismatches > 0 {
            return NotificationReliabilityReport(level: .needsRepair, mismatchCount: mismatches, shouldOfferRepair: true)
        }
        if input.pendingCount == 0 {
            return NotificationReliabilityReport(level: .degraded, mismatchCount: 0, shouldOfferRepair: false)
        }
        return NotificationReliabilityReport(level: .healthy, mismatchCount: 0, shouldOfferRepair: false)
    }

    private static func mismatchCount(_ input: NotificationReliabilityInput) -> Int {
        input.pendingOnlyCount + input.shadowOnlyCount + input.shadowErrorCount
    }
}

File: Android/app/src/main/java/com/example/supplementtracker/service/NotificationReliability.kt
package com.example.supplementtracker.service

enum class NotificationReliabilityLevel {
    HEALTHY,
    DEGRADED,
    NEEDS_REPAIR,
    INACTIVE
}

data class NotificationReliabilityInput(
    val permissionGranted: Boolean,
    val enabledByUser: Boolean,
    val hasActiveClient: Boolean,
    val activeSupplementCount: Int,
    val exactAlarmAvailable: Boolean,
    val batteryOptimizationIgnored: Boolean,
    val scheduledCount: Int,
    val missingPendingIntentCount: Int,
    val staleEntryCount: Int
)

data class NotificationReliabilityReport(
    val level: NotificationReliabilityLevel,
    val missingPendingIntentCount: Int,
    val staleEntryCount: Int,
    val shouldOfferRepair: Boolean
)

object NotificationReliabilityEvaluator {
    fun evaluate(input: NotificationReliabilityInput): NotificationReliabilityReport {
        val inactive = !input.enabledByUser || !input.permissionGranted || !input.hasActiveClient
        if (inactive || input.activeSupplementCount == 0) {
            return NotificationReliabilityReport(
                level = NotificationReliabilityLevel.INACTIVE,
                missingPendingIntentCount = input.missingPendingIntentCount,
                staleEntryCount = input.staleEntryCount,
                shouldOfferRepair = false
            )
        }
        val hasBrokenSchedule = input.missingPendingIntentCount > 0 || input.staleEntryCount > 0
        val noFutureSchedule = input.scheduledCount == 0
        val level = when {
            hasBrokenSchedule -> NotificationReliabilityLevel.NEEDS_REPAIR
            noFutureSchedule || !input.exactAlarmAvailable || !input.batteryOptimizationIgnored -> {
                NotificationReliabilityLevel.DEGRADED
            }
            else -> NotificationReliabilityLevel.HEALTHY
        }
        return NotificationReliabilityReport(
            level = level,
            missingPendingIntentCount = input.missingPendingIntentCount,
            staleEntryCount = input.staleEntryCount,
            shouldOfferRepair = level == NotificationReliabilityLevel.NEEDS_REPAIR
        )
    }
}

data class NotificationAlarmAudit(
    val scheduledCount: Int,
    val missingPendingIntentCount: Int,
    val staleEntryCount: Int
)
